### DISCUSSION TITLE
Rotation Matrix Clarification

### DISCUSSION THREAD

**Discussion by willgraham345** [bitcraze/discussions#1309] [Q&A]
I'm reaching out to see if I can get additional clarification on the rotation matrix within the `LighthouseBsGeometry` class shown in this [example](https://github.com/bitcraze/crazyflie-lib-python/blob/master/examples/lighthouse/write_lighthouse_mem.py). My system requires that I manually write values for the `origin` and `rotation_matrix` values. I want to clarify that I'm understanding the rotation matrix stored here to be the rotation matrix of R_{bs} in the formula found on the [Lighthouse Kalman Measurement Model](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/lighthouse/kalman_measurement_model/)
![image](https://github.com/bitcraze/crazyflie-lib-python/assets/49169367/88ab6670-a31f-4f31-8e6b-43203638a8e7)

Is that rotation matrix the changing it to be the rotation from global TO the new lighthouse position? 

As an example, say I take the Lighthouse and offset it 3m in the x-direction from the global origin (see picture below)
![bitcraze upload](https://github.com/bitcraze/crazyflie-lib-python/assets/49169367/473b4ea2-d7ba-419a-83ec-7e38e9a988d2)

If I take the lighthouse basestation frame as is (x is pointing through the glass, y to the left, z away from the screw) and rotate it so that the basestation is laying flat on its back, what would that rotation matrix be?

My guess is that the new rotation matrix should be a rotation of $-\pi/2$ about the y axis, meaning I need a rotation matrix of:
![image](https://github.com/bitcraze/crazyflie-lib-python/assets/49169367/49744d9f-ff58-49c6-92b2-36346c62e207)

which evaluates to:
![image](https://github.com/bitcraze/crazyflie-lib-python/assets/49169367/d4dc415f-d846-4b7e-b34e-15efe257a34d)
I'd then write that matrix to the `rotation_matrix` value within the `LighthouseBsGeometry`. Am I understanding this correctly, or is there more going on behind the scenes?

**Comment by knmcguire** [comment]
I think that you are correct here but the lighthouse geometry estimation itself is quite challenging to go through.

You have found this page as well with definitions and coordinate systems as well right? https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/lighthouse/terminology_definitions/#geometry-data

**Comment by willgraham345** [comment]
Yes I did see that link with the definitions, and my hope in making this discussion was to find the software equivalent of those formulas. I agree with the sentiment that the lighthouse geometry estimation methods are a challenge to get through. In an attempt to understand, I have made a few class diagrams and sequencing diagram trying to match these methods together. I'm hoping that I can find extra clarification on matching different formulas with the software in the Lighthouse estimations. 

Does that make more sense?

**Reply by knmcguire** [reply]
Ah yes perfectly. Like I said, I do think that you grabbed the concept of it of what the geometry estimation is about. How exactly the estimation is done is perhaps another topic for another thread :)
